flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferConsumer.java

```diff
@@ -163,6 +170,10 @@ public int getCached() {
 			return PositionMarker.getAbsolute(cachedPosition);
 		}
 
+		private int getLatest() {
+			return PositionMarker.getAbsolute(positionMarker.get());
```

Comment:
This is adding new synchronisation point (`volatile` read) and will probably decrease performance in low latency or huge clusters high throughput scenarios.

Let me think a bit about this issue.

Reply:
I ran some benchmark cases with 1ms bufferTimeout and no evident performance difference was seen before and after the fix. The followings are the result.

Before the fix:
Benchmark                 (channelsFlushTimeout)   Mode  Cnt      Score        Error         Units
NetworkThroughput          1000,1ms                     thrpt  30  23032.384 ± 871.883  ops/ms
KeyByBenchmarks.arrayKeyBy1MS                      thrpt   30    1923.863   ± 78.518   ops/ms
KeyByBenchmarks.tupleKeyBy1MS                      thrpt   30    3377.401   ± 216.982  ops/ms
InputBenchmark.mapRebalanceMapSink1MS     thrpt   30    6091.213   ±  92.658   ops/ms
InputBenchmark.mapSinkBufferTimeout1MS      thrpt   30    9107.194    ± 211.169   ops/ms

After the fix:
Benchmark          (channelsFlushTimeout)      Mode  Cnt      Score             Error      Units
NetworkThroughput 1000,1ms                          thrpt   30    23985.588 ± 990.037  ops/ms
KeyByBenchmarks.arrayKeyBy1MS                  thrpt   30      2011.356   ± 40.347    ops/ms
KeyByBenchmarks.tupleKeyBy1MS                  thrpt   30     3440.238   ± 211.906   ops/ms
InputBenchmark.mapRebalanceMapSink1MS thrpt   30      6118.888    ±  94.517   ops/ms 
InputBenchmark.mapSinkBufferTimeout1MS  thrpt   30      9120.144   ± 252.023   ops/ms

As discussed in the mailing list. High data skew between large number of channels (subpartitions) with low bufferTimeout can lead to throughput regression. The results are as follows.

Before the fix:
Benchmark        (channelsFlushTimeout)     Mode  Cnt    Score           Error       Units
NetworkThroughput     1000,1ms                   thrpt    30 18240.197 ± 1892.419 ops/ms

After the fix:
Benchmark         (channelsFlushTimeout)    Mode  Cnt     Score            Error    Units
NetworkThroughput   1000,1ms                     thrpt   30  24532.313 ± 1118.312 ops/ms